internal/auth/aws/tiktok_provider.go

```diff
@@ -0,0 +1,182 @@
+package aws
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/goccy/go-json"
+	"github.com/pkg/errors"
+)
+
+// TikTokConfig stores the client key and secret for TikTok API
+type TikTokConfig struct {
+	ClientKey    string
+	ClientSecret string
+}
+
+// TikTokProvider is a custom AWS credentials provider that uses the TikTok
+// API to obtain temporary credentials. It implements the aws.CredentialsProvider
+// interface and can be used with the AWS SDK for Go v2.
+type TikTokProvider struct {
+	HTTPClient *http.Client
+	Config     TikTokConfig
+	Resource   ResourceType
+}
+
+var _ aws.CredentialsProvider = (*TikTokProvider)(nil)
+
+// NewTikTokProvider creates a new TikTokProvider with the given client key,
+// client secret, and resource type. The resource type determines the type of
+// credentials to be retrieved. The provider uses the default HTTP client
+// for making requests to the TikTok API.
+func NewTikTokProvider(key, secret string, resourceType ResourceType) *TikTokProvider {
+	return &TikTokProvider{
+		HTTPClient: http.DefaultClient,
+		Config: TikTokConfig{
+			ClientKey:    key,
+			ClientSecret: secret,
+		},
+		Resource: resourceType,
+	}
+}
+
+func (p *TikTokProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	token, err := p.getClientToken()
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	switch p.Resource {
+	case NoneResourceType:
+		return aws.Credentials{
+			AccessKeyID:     p.Config.ClientKey,
+			SecretAccessKey: p.Config.ClientSecret,
+			SessionToken:    token,
+			Source:          "tiktok",
+		}, nil
+	case S3ResourceType:
+		time.Sleep(time.Duration(rand.Intn(5)+5) * time.Second)
```

Comment:
what is the purpose of time.Sleep here? If necessary, please add a comment

Reply:
it's used for waiting between getting client credentials and session, added a comment